File: app/src/main/java/de/bahnhoefe/deutschlands/bahnhofsfotos/OutboxActivity.java
package de.bahnhoefe.deutschlands.bahnhofsfotos;

import static java.util.stream.Collectors.toList;

import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.ContextThemeWrapper;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import java.util.List;

import de.bahnhoefe.deutschlands.bahnhofsfotos.databinding.ActivityOutboxBinding;
import de.bahnhoefe.deutschlands.bahnhofsfotos.db.DbAdapter;
import de.bahnhoefe.deutschlands.bahnhofsfotos.db.OutboxAdapter;
import de.bahnhoefe.deutschlands.bahnhofsfotos.dialogs.SimpleDialogs;
import de.bahnhoefe.deutschlands.bahnhofsfotos.model.InboxStateQuery;
import de.bahnhoefe.deutschlands.bahnhofsfotos.model.Upload;
import de.bahnhoefe.deutschlands.bahnhofsfotos.util.FileUtils;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class OutboxActivity extends AppCompatActivity {

    private static final String TAG = OutboxActivity.class.getSimpleName();

    private OutboxAdapter adapter;
    private DbAdapter dbAdapter;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        var baseApplication = (BaseApplication) getApplication();
        dbAdapter = baseApplication.getDbAdapter();

        var binding = ActivityOutboxBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        adapter = new OutboxAdapter(OutboxActivity.this, dbAdapter.getOutbox());
        binding.lstUploads.setAdapter(adapter);

        // item click
        binding.lstUploads.setOnItemClickListener((parent, view, position, id) -> {
            var upload = dbAdapter.getUploadById(id);
            Intent intent;
            if (upload.isProblemReport()) {
                intent = new Intent(OutboxActivity.this, ProblemReportActivity.class);
                intent.putExtra(ProblemReportActivity.EXTRA_UPLOAD, upload);
            } else {
                intent = new Intent(OutboxActivity.this, UploadActivity.class);
                intent.putExtra(UploadActivity.EXTRA_UPLOAD, upload);
            }
            startActivity(intent);
        });

        binding.lstUploads.setOnItemLongClickListener((parent, view, position, id) -> {
            var uploadId = String.valueOf(id);
            SimpleDialogs.confirmOkCancel(OutboxActivity.this, getResources().getString(R.string.delete_upload, uploadId), (dialog, which) -> {
                dbAdapter.deleteUpload(id);
                FileUtils.deleteQuietly(FileUtils.getStoredMediaFile(this, id));
                adapter.changeCursor(dbAdapter.getOutbox());
            });
            return true;
        });

        var query = dbAdapter.getPendingUploads(true).stream()
                .map(upload -> InboxStateQuery.builder().id(upload.getRemoteId()).build())
                .collect(toList());

        baseApplication.getRsapiClient().queryUploadState(query).enqueue(new Callback<>() {
            @Override
            public void onResponse(@NonNull Call<List<InboxStateQuery>> call, @NonNull Response<List<InboxStateQuery>> response) {
                if (response.isSuccessful()) {
                    var stateQueries = response.body();
                    if (stateQueries != null) {
                        dbAdapter.updateUploadStates(stateQueries);
                        adapter.changeCursor(dbAdapter.getOutbox());
                    }
                } else if (response.code() == 401) {
                    baseApplication.setAccessToken(null);
                    baseApplication.getRsapiClient().clearToken();
                    Toast.makeText(OutboxActivity.this, R.string.authorization_failed, Toast.LENGTH_LONG).show();
                    startActivity(new Intent(OutboxActivity.this, MyDataActivity.class));
                    finish();
                } else {
                    Log.w(TAG, "Upload states not processable");
                }
            }

            @Override
            public void onFailure(@NonNull Call<List<InboxStateQuery>> call, @NonNull Throwable t) {
                Log.e(TAG, "Error retrieving upload state", t);
                Toast.makeText(OutboxActivity.this,
                        R.string.error_retrieving_upload_state,
                        Toast.LENGTH_LONG).show();
            }
        });
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        menu.clear();
        getMenuInflater().inflate(R.menu.outbox, menu);
        return super.onPrepareOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.nav_delete_processed_uploads) {
            deleteCompletedUploads();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void deleteCompletedUploads() {
        var uploads = dbAdapter.getCompletedUploads();
        if (uploads.isEmpty()) {
            return;
        }

        new AlertDialog.Builder(new ContextThemeWrapper(this, R.style.AlertDialogCustom))
                .setIcon(R.mipmap.ic_launcher)
                .setTitle(R.string.confirm_delete_processed_uploads)
                .setPositiveButton(R.string.button_ok_text, (dialog, which) -> {
                    for (Upload upload : uploads) {
                        dbAdapter.deleteUpload(upload.getId());
                        FileUtils.deleteQuietly(FileUtils.getStoredMediaFile(this, upload.getId()));
                    }
                    adapter.changeCursor(dbAdapter.getOutbox());
                })
                .setNegativeButton(R.string.button_cancel_text, null)
                .create().show();
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        adapter.changeCursor(dbAdapter.getOutbox());
    }

}

File: app/src/main/kotlin/de/bahnhoefe/deutschlands/bahnhofsfotos/OutboxActivity.kt
package de.bahnhoefe.deutschlands.bahnhofsfotos

import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.ContextThemeWrapper
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup.MarginLayoutParams
import android.widget.AdapterView
import android.widget.AdapterView.OnItemClickListener
import android.widget.AdapterView.OnItemLongClickListener
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.updateLayoutParams
import dagger.hilt.android.AndroidEntryPoint
import de.bahnhoefe.deutschlands.bahnhofsfotos.databinding.ActivityOutboxBinding
import de.bahnhoefe.deutschlands.bahnhofsfotos.db.DbAdapter
import de.bahnhoefe.deutschlands.bahnhofsfotos.db.OutboxAdapter
import de.bahnhoefe.deutschlands.bahnhofsfotos.dialogs.SimpleDialogs
import de.bahnhoefe.deutschlands.bahnhofsfotos.model.InboxStateQuery
import de.bahnhoefe.deutschlands.bahnhofsfotos.model.Upload
import de.bahnhoefe.deutschlands.bahnhofsfotos.rsapi.RSAPIClient
import de.bahnhoefe.deutschlands.bahnhofsfotos.util.FileUtils
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import javax.inject.Inject

private val TAG = OutboxActivity::class.java.simpleName

@AndroidEntryPoint
class OutboxActivity : AppCompatActivity() {
    private lateinit var adapter: OutboxAdapter

    @Inject
    lateinit var dbAdapter: DbAdapter

    @Inject
    lateinit var rsapiClient: RSAPIClient

    public override fun onCreate(savedInstanceState: Bundle?) {
        enableEdgeToEdge()
        super.onCreate(savedInstanceState)
        val binding = ActivityOutboxBinding.inflate(layoutInflater)
        ViewCompat.setOnApplyWindowInsetsListener(binding.getRoot()) { v, windowInsets ->
            val insets = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.updateLayoutParams<MarginLayoutParams> {
                bottomMargin = insets.bottom
                topMargin = insets.top
            }
            WindowInsetsCompat.CONSUMED
        }
        setContentView(binding.root)

        if (!rsapiClient.isLoggedIn) {
            Toast.makeText(this, R.string.please_login, Toast.LENGTH_LONG).show()
            startActivity(Intent(this@OutboxActivity, MyDataActivity::class.java))
            finish()
            return
        }

        adapter = OutboxAdapter(this@OutboxActivity, dbAdapter.outbox)
        binding.lstUploads.adapter = adapter

        binding.lstUploads.onItemClickListener =
            OnItemClickListener { _: AdapterView<*>?, _: View?, _: Int, id: Long ->
                val upload = dbAdapter.getUploadById(id)
                val intent: Intent
                if (upload!!.isProblemReport) {
                    intent = Intent(this@OutboxActivity, ProblemReportActivity::class.java)
                    intent.putExtra(EXTRA_PROBLEM_UPLOAD, upload)
                } else {
                    intent = Intent(this@OutboxActivity, UploadActivity::class.java)
                    intent.putExtra(EXTRA_UPLOAD_UPLOAD, upload)
                }
                startActivity(intent)
            }
        binding.lstUploads.onItemLongClickListener =
            OnItemLongClickListener { _: AdapterView<*>?, _: View?, _: Int, id: Long ->
                dbAdapter.getUploadById(id)?.let { deleteUploadEntry(it) }
                true
            }

        val query = dbAdapter.getPendingUploads(true)
            .map { upload: Upload? ->
                InboxStateQuery(
                    upload!!.remoteId
                )
            }

        rsapiClient.queryUploadState(query)
            .enqueue(object : Callback<List<InboxStateQuery>> {
                override fun onResponse(
                    call: Call<List<InboxStateQuery>>,
                    response: Response<List<InboxStateQuery>>
                ) {
                    if (response.isSuccessful) {
                        val stateQueries = response.body()
                        if (stateQueries != null) {
                            dbAdapter.updateUploadStates(stateQueries)
                            adapter.changeCursor(dbAdapter.outbox)
                        }
                    } else if (response.code() == 401) {
                        rsapiClient.clearToken()
                        Toast.makeText(
                            this@OutboxActivity,
                            R.string.authorization_failed,
                            Toast.LENGTH_LONG
                        ).show()
                        startActivity(Intent(this@OutboxActivity, MyDataActivity::class.java))
                        finish()
                    } else {
                        Log.w(TAG, "Upload states not processable")
                    }
                }

                override fun onFailure(call: Call<List<InboxStateQuery>>, t: Throwable) {
                    Log.e(TAG, "Error retrieving upload state", t)
                    Toast.makeText(
                        this@OutboxActivity,
                        R.string.error_retrieving_upload_state,
                        Toast.LENGTH_LONG
                    ).show()
                }
            })
    }

    private fun deleteUploadEntry(upload: Upload) {
        val remoteDelete = (upload.remoteId != null && upload.uploadState.isPending)
        val message =
            if (remoteDelete) {
                getString(R.string.delete_outbox_entry_remote, upload.remoteId)
            } else {
                getString(R.string.delete_outbox_entry_local, upload.id)
            }

        SimpleDialogs.confirmOkCancel(this@OutboxActivity, message) { _: DialogInterface?, _: Int ->
            if (remoteDelete) {
                rsapiClient.deleteInboxEntry(upload.remoteId!!).enqueue(object : Callback<Void> {
                    override fun onResponse(call: Call<Void>, response: Response<Void>) {
                        if (response.isSuccessful) {
                            deleteUploadEntryLocally(upload.id!!)
                        } else if (response.code() == 401) {
                            rsapiClient.clearToken()
                            Toast.makeText(
                                this@OutboxActivity,
                                R.string.authorization_failed,
                                Toast.LENGTH_LONG
                            ).show()
                        } else {
                            Toast.makeText(
                                this@OutboxActivity,
                                getString(
                                    R.string.error_deleting_inbox_entry,
                                    "${response.code()}: ${response.message()}"
                                ),
                                Toast.LENGTH_LONG
                            ).show()
                        }
                    }

                    override fun onFailure(call: Call<Void>, t: Throwable) {
                        Log.e(TAG, "Error deleting remote inbox entry", t)
                        Toast.makeText(
                            this@OutboxActivity,
                            getString(R.string.error_deleting_inbox_entry, t.message),
                            Toast.LENGTH_LONG
                        ).show()
                    }

                })
            } else {
                deleteUploadEntryLocally(upload.id!!)
            }
        }
    }

    private fun deleteUploadEntryLocally(id: Long) {
        dbAdapter.deleteUpload(id)
        FileUtils.deleteQuietly(FileUtils.getStoredMediaFile(this, id))
        adapter.changeCursor(dbAdapter.outbox)
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        menu.clear()
        menuInflater.inflate(R.menu.outbox, menu)
        return super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.nav_delete_processed_uploads) {
            deleteCompletedUploads()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    private fun deleteCompletedUploads() {
        val uploads = dbAdapter.completedUploads
        if (uploads.isEmpty()) {
            return
        }
        AlertDialog.Builder(ContextThemeWrapper(this, R.style.AlertDialogCustom))
            .setIcon(R.mipmap.ic_launcher)
            .setTitle(R.string.confirm_delete_processed_uploads)
            .setPositiveButton(R.string.button_ok_text) { _: DialogInterface?, _: Int ->
                for (upload in uploads) {
                    dbAdapter.deleteUpload(upload.id!!)
                    FileUtils.deleteQuietly(FileUtils.getStoredMediaFile(this, upload.id))
                }
                adapter.changeCursor(dbAdapter.outbox)
            }
            .setNegativeButton(R.string.button_cancel_text, null)
            .create().show()
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        adapter.changeCursor(dbAdapter.outbox)
    }

}